### DISCUSSION TITLE
Documentation update for managed python environment?

### DISCUSSION THREAD

**Discussion by aairfccha** [radioactivedecay/radioactivedecay#111] [General]
I just found out that Ubuntu has turned python into an externally managed environment. The basic idea is to prevent the user from breaking the system by removing or overwriting system-essential python packages. Unfortunately, this breaks the usual easy way to install python packages. I guess the best intended way is a virtual environment (venv), the sledgehammer approach is the parameter "--break-system-packages" but having little experience with python this really is just a guess. A third option is apparently pipx, but I have even less of an idea what that actually does.

https://peps.python.org/pep-0668/
https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/
https://python.land/virtual-environments/pipx

**Comment by php1ic** [comment]
The link to the PEP page suggests that `pip install --user` is blocked for the same reason, but if you follow the link to the [up-to date spec](https://packaging.python.org/en/latest/specifications/externally-managed-environments/#externally-managed-environments), it does not seem as clear, but I believe is still the case.

I do not have the latest version of Ubuntu installed, are you able to test if `pip install --user` works?

**Reply by aairfccha** [reply]
At least for an installation of radioactivedecay originally done the regular way before the change to a managed environment, `pip install --user -U radioactivedecay` doesn't work.

**Comment by alexmalins** [comment]
> I guess the best intended way is a virtual environment (venv)

The first answer [here](https://stackoverflow.com/questions/75608323/how-do-i-solve-error-externally-managed-environment-every-time-i-use-pip-3) suggests the solution is as you wrote above - install radioactivedecay into some type of virtual environment.

There's many options for managing python virtual environments: Python's inbuilt venv, virtualenv, pipx, conda, poetry, uv, pdm,...
We already maintain a conda-forge feedstock and document the conda option in the docs. Maybe you could suggest some documentation text to let users know that installing in a virtual env of some sort may be required on your system?

It's unlikely we'd be able to get radioactivedecay available on the default Ubuntu/Debian apt servers so you could install via apt like `apt install python3-radioactivedecay`. And I wouldn't recommend that anyway as if one of radioactivedecay's dependency requirements conflicts with whatever version of Ubuntu needs, there's a problem.

**Reply by aairfccha** [reply]
Here is a quick and dirty setup guide for a virtual environment with virtualenv. Initial condition: Linux with managed python 3 installation and working pip.

First step: install python package virtualenv. In the spirit of a managed environment you should probably do this without the sledgehammer --break-system-packages, so:
`apt install python3-virtualenv`
Then move the terminal to a suitable location for the folder holding the virtual environment. Here I simply (ab)using "Documents"
`cd Documents`
Now create the virtual environment. Here it is called "name" but if you plan to use several, a more descriptive name might be a good idea, especially if you want to keep them in the same folder.
`virtualenv name`
Now you have a folder "name" in your folder "Documents". To work within this virtual environment, just type
`source name/bin/activate`
Now you should be able to install and uninstall any package with the usual pip command without messing up the entire system.

To activate the virtual environment again after a reboot, navigate the terminal to the location of the virtual environment folder
`cd Documents`
And of course
`source name/bin/activate`

One caveat I found so far: Within the virtual environment, you can run .py files just fine, but IDLE apparently opens in the "global" python environment by default anyway. For IDLE in the virtual environment, you activate the environment and type
`python -m idlelib.idle`
I saw a mention that you can set an alias for that, but so far I haven't bothered to go in that deep.

**Reply by Cs137** [reply]
Since people like to know how to exit things, I'd add a sentence like that:

> To deactivate the virtual environment use the `deactivate` command.
